File: go/protein-translation/protein_translation.go
package protein

import (
	"errors"
)

// ErrStop indicates a STOP codon was reached
var ErrStop = errors.New("ErrStop")

// ErrInvalidBase indicates an invalid codon
var ErrInvalidBase = errors.New("ErrInvalidBase")

// FromCodon takes a codon and returns a protein
func FromCodon(codon string) (string, error) {

	switch codon {
	case "AUG":
		return "Methionine", nil
	case "UUU", "UUC":
		return "Phenylalanine", nil
	case "UUA", "UUG":
		return "Leucine", nil
	case "UCU", "UCC", "UCA", "UCG":
		return "Serine", nil
	case "UAU", "UAC":
		return "Tyrosine", nil
	case "UGU", "UGC":
		return "Cysteine", nil
	case "UGG":
		return "Tryptophan", nil
	case "UAA", "UAG", "UGA":
		return "", ErrStop
	default:
		return "", ErrInvalidBase
	}
}

// FromRNA takes an rna a returns a list of proteins
func FromRNA(rna string) ([]string, error) {

	var proteins []string
	var processingError error

	for i := 0; i <= len(rna)-3; i += 3 {
		protein, err := FromCodon(rna[i : i+3])
		if err == ErrStop {
			break
		}
		if err == ErrInvalidBase {
			processingError = ErrInvalidBase
			break
		}
		proteins = append(proteins, protein)
	}
	return proteins, processingError
}

File: lua/protein-translation/protein-translation.lua
local function translate_codon(codon)
    if codon == "AUG" then
        return "Methionine"
    elseif codon == "UGG" then
        return "Tryptophan"
    elseif codon == "UUU" or codon == "UUC" then
        return "Phenylalanine"
    elseif codon == "UUA" or codon == "UUG" then
        return "Leucine"
    elseif codon == "UAU" or codon == "UAC" then
        return "Tyrosine"
    elseif codon == "UGU" or codon == "UGC" then
        return "Cysteine"
    elseif codon == "UAA" or codon == "UAG" or codon == "UGA" then
        return "STOP"
    elseif codon == "UCU" or codon == "UCC" or codon == "UCA" or codon == "UCG" then
        return "Serine"
    else
        error("unknown codon: " .. codon)
    end
end

local function translate_rna_strand(rna_strand)
    local proteins = {}

    for codon in rna_strand:gmatch("...") do
        local protein = translate_codon(codon)
        if protein == "STOP" then
            return proteins
        end
        table.insert(proteins, protein)
    end

    return proteins
end

return {
    codon = translate_codon,
    rna_strand = translate_rna_strand
}

